File: sources/CSharp/src/Ers/SubModel/Component/BasicRenderComponent.cs
using System;
using System.Numerics;
using System.Runtime.InteropServices;
using Ers.Engine;

namespace Ers
{
    /// <summary>
    /// A component for a quick rendering setup.
    ///
    /// <para>
    /// Apply just a color to render a rectangle in 2D and a cube in 3D.
    /// Add an <see cref="Ers.InstancedModel"/> to set a custom 3D model.
    /// </para>
    /// </summary>
    [StructLayout(LayoutKind.Sequential)]
    public struct BasicRenderComponent : ICoreComponent
    {
        /// <summary>
        /// The color of the entity.
        /// </summary>
        public Color Color
        {
            get => Color.FromInt(ErsEngine.ERS_BasicRenderComponent_GetColor(CorePointer()));
            set => ErsEngine.ERS_BasicRenderComponent_SetColor(CorePointer(), value.Value);
        }

        /// <summary>
        /// The <see cref="Ers.InstancedModel"/> used for 3D rendering.
        /// </summary>
        public InstancedModel InstancedModel
        {
            get {
                unsafe
                {
                    IntPtr ptr = ErsEngine.ERS_BasicRenderComponent_GetInstancedModel(CorePointer());
                    return new InstancedModel(ptr);
                }
            }
            set {
                ErsEngine.ERS_BasicRenderComponent_SetInstancedModel(CorePointer(), value.Data);
            }
        }

        /// <summary>
        /// The type ID of the component in the ERS core.
        /// </summary>
        /// <returns></returns>
        public static nuint CoreTypeId() => ErsEngine.ERS_BasicRenderComponent_TypeId();

        private IntPtr CorePointer()
        {
            unsafe
            {
                fixed(BasicRenderComponent* ptr = &this)
                {
                    return (IntPtr)ptr;
                }
            }
        }
    }
}

File: sources/CPP/include/Ers/SubModel/Component/BasicRenderComponent.cpp
#include "BasicRenderComponent.h"
#include "Ers/Api.h"
#include "Ers/Visualization/InstancedModel.h"

namespace Ers
{
    uint32_t BasicRenderComponent::CoreTypeId()
    {
        return ersAPIFunctionPointers.ERS_BasicRenderComponent_TypeId();
    }

    Color BasicRenderComponent::GetColor() const
    {
        return Color::FromInt(ersAPIFunctionPointers.ERS_BasicRenderComponent_GetColor(const_cast<BasicRenderComponent*>(this)));
    }

    void BasicRenderComponent::SetColor(Ers::Color color)
    {
        ersAPIFunctionPointers.ERS_BasicRenderComponent_SetColor(this, color.Value);
    }

    Ers::InstancedModel* BasicRenderComponent::GetInstancedModel()
    {
        void* ptr = ersAPIFunctionPointers.ERS_BasicRenderComponent_GetInstancedModel(this);
        return static_cast<Ers::InstancedModel*>(ptr);
    }

    void BasicRenderComponent::SetInstancedModel(Ers::InstancedModel* model)
    {
        ersAPIFunctionPointers.ERS_BasicRenderComponent_SetInstancedModel(this, model);
    }
} // namespace Ers
